File: src/widgets/label.c
/* Claro Graphics - an abstraction layer for native UI libraries
 * 
 * $Id$
 * 
 * The contents of this file are subject to the Mozilla Public License
 * Version 1.1 (the "License"); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 * http://www.mozilla.org/MPL/
 * 
 * Software distributed under the License is distributed on an "AS IS"
 * basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See the
 * License for the specific language governing rights and limitations
 * under the License.
 * 
 * See the LICENSE file for more details.
 */


#include "graphics.h"
#include "platform.h"

object_t *label_widget_create( object_t *parent, bounds_t *bounds, int flags )
{
	object_t *label;
	
	assert_valid_widget( parent, "parent" );
	
	label = default_widget_create(parent, sizeof(label_widget_t), 
										   "claro.graphics.widgets.label", bounds, 
										   flags, cgraphics_label_widget_create);
	
	label_set_justify( label, flags );
	label_set_text( label, "" );
	
	return label;
}

object_t *label_widget_create_with_text( object_t *parent, bounds_t *bounds, int flags, const char *text )
{
	object_t *label;
	
	assert_valid_widget( parent, "parent" );
	
	label = label_widget_create( parent, bounds, flags );
	
	label_set_text( label, text );
	
	return label;
}

void label_set_justify( object_t *obj, int flags )
{
	WIDGET(obj)->flags = flags;
	
	assert_valid_label_widget( obj, "obj" );
	
	cgraphics_label_update_justify( (widget_t *)obj );
}

void label_set_text( object_t *obj, const char *text )
{
	label_widget_t *w = (label_widget_t *)obj;
	
	assert_valid_label_widget( obj, "obj" );
	
	strscpy( w->text, text, CLARO_LABEL_MAXIMUM );
	
	cgraphics_label_update_text( (widget_t *)w );
}

File: src/platform/cocoa/label.m
/* Claro Graphics - an abstraction layer for native UI libraries
 * 
 * $Id$
 * 
 * The contents of this file are subject to the Mozilla Public License
 * Version 1.1 (the "License"); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 * http://www.mozilla.org/MPL/
 * 
 * Software distributed under the License is distributed on an "AS IS"
 * basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See the
 * License for the specific language governing rights and limitations
 * under the License.
 * 
 * See the LICENSE file for more details.
 */


#define _PLATFORM_INC
#include "graphics.h"
#include "platform/macosx_cocoa.h"

/* Label */

/* ClaroLabel (subclassed from NSTextField) */
@interface ClaroLabel : NSTextField
{
	widget_t *cw;
}

/* internal init function */
- (void)setClaroWidget:(widget_t *)widget;

/* notification responders */
- (void)claroResize:(NSNotification *)aNotification;
- (void)claroMove:(NSNotification *)aNotification;
- (void)claroClose:(NSNotification *)aNotification;
@end

@implementation ClaroLabel

- (void)claroResize:(NSNotification *)aNotification
{
	NSRect frame = [self frame];
	widget_set_size( cw, frame.size.width, frame.size.height, 1 );
}

- (void)claroMove:(NSNotification *)aNotification
{
	NSRect frame = [self frame];
	widget_set_position( cw, frame.origin.x, frame.origin.y, 1 );
}

- (void)claroClose:(NSNotification *)aNotification
{
	widget_destroy( cw );
}

- (void)setClaroWidget:(widget_t *)widget
{
	cw = widget;
	
	[[NSNotificationCenter defaultCenter] addObserver:self
		selector:@selector(claroClose:) name:NSWindowWillCloseNotification
		object:self];
	
	[[NSNotificationCenter defaultCenter] addObserver:self
		selector:@selector(claroResize:) name:NSWindowDidResizeNotification
		object:self];
	
	[[NSNotificationCenter defaultCenter] addObserver:self
		selector:@selector(claroMove:) name:NSWindowDidMoveNotification
		object:self];
}

@end

void cgraphics_label_widget_create(widget_t *widget)
{
	ClaroLabel *tb;
	NSView *parent = cgraphics_get_native_parent( widget );
	
	if ( widget->size_req->h == -1 )
		widget->size_req->h = 21;
	
	tb = [[ClaroLabel alloc]
		initWithFrame: NSMakeRect( widget->size_req->x, widget->size_req->y, widget->size_req->w, widget->size_req->h )
		];
	
	[tb setEditable: NO];
	[tb setAlignment: NSLeftTextAlignment];
	[tb setDrawsBackground: NO];
	[tb setBordered: NO];
	
	[tb setClaroWidget:widget];
	[parent addSubview: tb];
	
	widget->native = (NSControl *)tb;
}

void cgraphics_label_update_text(textbox_widget_t *widget)
{
	ClaroLabel *tb = (ClaroLabel *)widget->widget.native;
	NSString *str = [[NSString alloc] initWithUTF8String: widget->text];
	[tb setStringValue:str];
}

void cgraphics_label_update_justify( widget_t *widget )
{
	ClaroLabel *tb = (ClaroLabel *)widget->native;
	
	switch ( widget->flags ) {
		case cLabelRight:
			[tb setAlignment: NSRightTextAlignment];
			break;
		case cLabelCenter:
			[tb setAlignment: NSCenterTextAlignment];
			break;
		case cLabelFill:
		case cLabelLeft:
		default:
			[tb setAlignment: NSLeftTextAlignment];
			break;
	}
}

